source/MapPanel.cpp

```diff
@@ -1386,10 +1402,50 @@ void MapPanel::DrawSystems()
 
 	// Draw the circles for the systems.
 	double zoom = Zoom();
+	BatchDrawList starBatch;
+
+	const float ringFade = mapIsStarry ? 1.5 - 1.25 * zoom : 1.;
 	for(const Node &node : nodes)
 	{
 		Point pos = zoom * (node.position + center);
-		RingShader::Draw(pos, OUTER, INNER, node.color);
+		if(!mapIsStarry)
+			RingShader::Draw(pos, OUTER, INNER, node.color);
+		else
+		{
+			// System rings fade as you zoom in if starry map is enabled.
+			const float alpha = max(ringFade, node.mapIcons.empty() ? .9f : 0.f);
+			RingShader::Draw(pos, OUTER, INNER, node.color.Additive(alpha));
+
+			// Ensures every multiple-star system has a characteristic, deterministic rotation.
+			Angle starAngle = 0;
+			Angle angularSpacing = 0;
+			Point starOffset = Point(0, 0);
+			int maxStars = 5;
+
+			if(node.mapIcons.size() > 1)
+			{
+				int starsToDraw = min(static_cast<int>(node.mapIcons.size()), maxStars);
+				starAngle = node.name.length() + node.position.Length();
+				angularSpacing = 360. / starsToDraw;
+				starOffset = starsToDraw * Point(2., 2.);
+			}
+
+			// Draw the star sprites
+			int starIncrement = 0;
+			for(const Sprite *star : node.mapIcons)
+			{
+				starIncrement += 1;
+				starAngle += angularSpacing;
+				const Body starBody(star, pos + zoom * starOffset * starAngle.Unit(),
+					Point(0, 0), 0, sqrt(zoom) / 2, min(zoom + 0.3, 0.9));
+				starBatch.Add(starBody);
+				if(starIncrement > maxStars)
+					break;
```

Comment:
`starIncrement` is a bad name. Try `starsDrawn` or similar.

Having the check on if the maximum number of stars have been drawn at the end of the loop also doesn't seem right.
If `maxStars` is 5 (as it currently is) this could end up drawing 6 stars:

`starIncrement` initialised to 0.
First iteration begins.
`starIncrement` incremented to 1.
Draw first star.
Check `starIncrement`. Not more than 5 so continue.
Second iteration begins.
`starIncrement` incremented to 2.
Draw second star.
Check `starIncrement`. Not more than 5 so continue.
Third iteration begins.
`starIncrement` incremented to 3.
Draw third star.
Check `starIncrement`. Not more than 5 so continue.
Fourth iteration begins.
`starIncrement` incremented to 4.
Draw fourth star.
Check `starIncrement`. Not more than 5 so continue.
Fifth iteration begins.
`starIncrement` incremented to 5.
Draw fifth star.
Check `starIncrement`. Not more than 5 so continue.
Sixth iteration begins.
`starIncrement` incremented to 6.
Draw sixth star.
Check `starIncrement`. More than 5 so stop.

Since the number stars that should be drawn is calculated earlier (`starsToDraw`), I suggest moving that variable out of the if statement it is currently in, and modifying this loop to be:
```suggestion
			for(int i = 0; i < starsToDraw; ++i)
			{
				starAngle += angularSpacing;
				const Sprite *star = node.mapIcons(i);
				const Body starBody(star, pos + zoom * starOffset * starAngle.Unit(),
					Point(0, 0), 0, sqrt(zoom) / 2, min(zoom + 0.3, 0.9));
				starBatch.Add(starBody);
```